python/text-to-speech/test_main.py

```diff
@@ -0,0 +1,317 @@
+        mock_response = MagicMock()
+        mock_response.text = "fake_access_token"
+        mock_response_request = MagicMock()
+        mock_response_request.content = DECODED_RESULT_AZURE
+        with unittest.mock.patch.object(
+            requests, "post", return_value=mock_response
+        ), unittest.mock.patch.object(
+            requests, "request", return_value=mock_response_request
+        ):
+            # Call the speech method.
+            got = instance.speech("hi", "en-US")
+        # Assert the result.
+        self.assertEqual(got, DECODED_RESULT_AZURE)
+
+    def test_speech_invalid_credential(self) -> None:
+        """Test credentials for speech method."""
+        instance = self.get_azure_instance("WRONG_API_KEY")
+        # Mock the requests.post method used in get_token.
+        mock_response = requests.Response
+        mock_response.text = "fake_access_token"
+        with unittest.mock.patch.object(
+            requests, "post", return_value=mock_response
+        ), unittest.mock.patch.object(
+            requests, "request", return_value=Exception("Error.")
+        ):
+            # Assert the raise.
+            self.assertRaises(Exception, instance.speech, "hi", "en-US")
+
+
+class AWSTest(unittest.TestCase):
+    """AWS API Test Cases."""
+
+    def get_aws_instance(self, key: str, secret_key: str) -> main.AWS:
+        """Set AWS instance with request."""
+        req = MyRequest({
+            "payload": {
+                "provider": "aws",
+                "text": "hi",
+                "language": "en-US",
+            },
+            "variables": {
+                "API_KEY": key,
+                "SECRET_API_KEY": secret_key,
+            }
+        })
+        return main.AWS(req)
+
+    @parameterized.expand([
+        # Missing API_KEY.
+        (None, "123"),
+        # Missing SECRET_API_KEY.
+        ("123", None),
+        # Missing Both.
+        (None, None),
+    ])
+    def test_validate_request(self, key: str, secret_key: str) -> None:
+        """Test 'validate_request' method when fields missing or invalid."""
+        self.assertRaises(ValueError, self.get_aws_instance, key, secret_key)
+
+    def test_speech(self) -> None:
+        """Test speech method for successful TextToSpeech synthesis."""
+        instance = self.get_aws_instance("123", "123")
+        # Set up mock.
+        raw_stream = io.BytesIO(DECODED_RESULT_AWS)
+        size = len(DECODED_RESULT_AWS)
+        stream_body_obj = botocore.response.StreamingBody(raw_stream, size)
+        with unittest.mock.patch.object(
+            boto3.Session, "client"
+        ) as mock_client:
+            mock_client.return_value.synthesize_speech.return_value = {
+                "AudioStream": stream_body_obj,
+                "ContentType": "bytes",
+                "RequestCharacters": 123,
+            }
+            got = instance.speech("hi", "en-US")
+            want = DECODED_RESULT_AWS
```

Comment:
Does this line need to be inside the `with` block?

Reply:
I think the instance.speech line needs to be inside the with block because I am getting errors about a security token being invalid if I run the tests. But the want can be moved outside the with block.